knowledge-base/grid-binding-to-expando-object.md

```diff
@@ -188,82 +185,395 @@ In addition to the sample below, there is a [complete runnable project in GitHub
+### Filtering
 
->caption Using ExpandoObject and LoadGroupsOnDemand
+* Filtering in nullable data requires you to use [`OnRead` data binding]({%slug common-features-data-binding-onread%}). Populate the missing `MemberType` properties in the `CompositeFilterDescriptor`s in the [`OnRead` event argument]({%slug common-features-data-binding-onread%}#event-argument), namely in `args.Request.Filters`. Also see the [related public item](https://feedback.telerik.com/blazor/1499221-incorrect-filter-operators-when-the-fieldtype-is-set-to-a-nullable-numeric-data-type). The following code snippet is part of the full example below.
 
-````CSHTML
-@using System.Dynamic
-@using Telerik.DataSource
-@using Telerik.DataSource.Extensions
+<div class="skip-repl"></div>
 
-<TelerikGrid OnRead="@OnGridRead"
-             TItem="@ExpandoObject"
-             Pageable="true"
-             Sortable="true"
-             FilterMode="@GridFilterMode.FilterRow"
-             Groupable="true"
-             LoadGroupsOnDemand="true"
-             OnStateInit="@OnGridStateInit">
-    <GridColumns>
-        @{
-            if (GridData != null && GridData.Any())
-            {
-                // The first data item should always contain non-null values for the property types to be determined.
-                // Use another data item or OnModelInit to populate default values for newly added items.
-                var firstDataItem = (IDictionary<string, object>)GridData.First();
+````CS
+private async Task OnGridRead(GridReadEventArgs args)
+{
+    args.Request.Filters.OfType<CompositeFilterDescriptor>()
+        .Each(x =>
+        {
+            x.FilterDescriptors.OfType<FilterDescriptor>()
+                .Each(y => y.MemberType = GetPropertyType(y.Member, true));
+        });
 
-                foreach (var item in firstDataItem)
+    // ...
+}
+````
+
+* [Row filtering]({%slug grid-filter-row%}) in nullable data requires [`FilterCellTemplate`]({%slug grid-templates-filter%}) for numeric and boolean properties. See columns `PropertyInt` and `PropertyBool` in the example below.
+
+* [`CheckboxList` filtering]({%slug grid-checklist-filter%}) in non-nullable data requires `OnRead` data binding and the same approach as in the first bullet item above.
+
+* The built-in `CheckboxList` filtering doesn't support nullable data. Instead, use the default [filter menu]({%slug grid-filter-menu%}) or [filter row]({%slug grid-filter-row%}). Using a filter menu template with a custom checkbox list is possible.
+
+* `CheckboxList` filtering in grouped data with `OnRead` event requires a [`FilterMenuTemplate`]({%slug grid-templates-filter%}). You can use a [`TelerikCheckBoxListFilter` component with custom data]({%slug grid-checklist-filter%}#custom-data). The following code snippet is part of the full example below. See column `PropertyBool`.
+
+<div class="skip-repl"></div>
```

Comment:
Again, I suggest indenting the code snippet and all other nested snippets that follow.